From 9720cbf8b2b6595f79e08f2964c53e47c96324f4 Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Wed, 30 Nov 2016 20:52:24 +0000
Subject: [PATCH] [CONFIGURATION-644] Fix for a duplicated header comment.

Under certain circumstances, the header comment managed by
PropertiesConfigurationLayout gets duplicated. This commit fixes
this problem.

Thanks to Andrew DeMaria for the patch.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/configuration/trunk@1772114 13f79535-47bb-0310-9956-ffa450edef68
---
 .../PropertiesConfigurationLayout.java        |  8 +++++---
 .../TestPropertiesConfigurationLayout.java    | 20 +++++++++++++++++++
 2 files changed, 25 insertions(+), 3 deletions(-)

diff --git a/src/main/java/org/apache/commons/configuration2/PropertiesConfigurationLayout.java b/src/main/java/org/apache/commons/configuration2/PropertiesConfigurationLayout.java
index 952fecce82..61a41d5e27 100644
--- a/src/main/java/org/apache/commons/configuration2/PropertiesConfigurationLayout.java
+++ b/src/main/java/org/apache/commons/configuration2/PropertiesConfigurationLayout.java
@@ -796,8 +796,7 @@ private String extractComment(List<String> commentLines, int from, int to)
      */
     private int checkHeaderComment(List<String> commentLines)
     {
-        if (loadCounter == 1 && getHeaderComment() == null
-                && layoutData.isEmpty())
+        if (loadCounter == 1 && layoutData.isEmpty())
         {
             // This is the first comment. Search for blanc lines.
             int index = commentLines.size() - 1;
@@ -806,7 +805,10 @@ private int checkHeaderComment(List<String> commentLines)
             {
                 index--;
             }
-            setHeaderComment(extractComment(commentLines, 0, index - 1));
+            if (getHeaderComment() == null)
+            {
+                setHeaderComment(extractComment(commentLines, 0, index - 1));
+            }
             return index + 1;
         }
         else
diff --git a/src/test/java/org/apache/commons/configuration2/TestPropertiesConfigurationLayout.java b/src/test/java/org/apache/commons/configuration2/TestPropertiesConfigurationLayout.java
index c90279e5d7..1916fb95ba 100644
--- a/src/test/java/org/apache/commons/configuration2/TestPropertiesConfigurationLayout.java
+++ b/src/test/java/org/apache/commons/configuration2/TestPropertiesConfigurationLayout.java
@@ -229,6 +229,26 @@ public void testHeaderCommentWithBlancs() throws ConfigurationException
         assertNull("Wrong comment for property", layout.getComment(TEST_KEY));
     }
 
+    /**
+     * Tests if a header comment containing blanc lines is correctly detected and doesn't overflow into the property
+     * comment in the case that the header comment is already set
+     */
+    @Test
+    public void testHeaderCommentWithBlancsAndPresetHeaderComment() throws ConfigurationException
+    {
+        String presetHeaderComment = "preset" + TEST_COMMENT + CRNORM + CRNORM + TEST_COMMENT;
+        builder.addComment(TEST_COMMENT);
+        builder.addComment(null);
+        builder.addComment(TEST_COMMENT);
+        builder.addComment(null);
+        builder.addProperty(TEST_KEY, TEST_VALUE);
+        layout.setHeaderComment(presetHeaderComment);
+        layout.load(config, builder.getReader());
+        assertEquals("Wrong header comment", presetHeaderComment,
+                     layout.getCanonicalHeaderComment(false));
+        assertNull("Wrong comment for property", layout.getComment(TEST_KEY));
+    }
+
     /**
      * Tests if a header comment is correctly detected when it contains blanc
      * lines and the first property has a comment, too.
